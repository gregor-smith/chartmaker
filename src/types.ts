import {
    Record as Record_,
    Literal,
    String,
    Union,
    InstanceOf,
    Boolean,
    Static,
    Number as Number_,
    Array as Array_,
    Runtype,
    Partial as Partial_
} from 'runtypes'

import {
    CHART_ALBUMS_COUNT,
    MAX_SCREENSHOT_SCALE,
    MAX_COLLAGE_ROWS_X,
    MAX_COLLAGE_ROWS_Y,
    STATE_VERSION,
    EXPORT_CHART_PLACEHOLDER
} from '@/constants'


const Integer = Number_.withConstraint(Number.isSafeInteger)


function IntegerRange(minimum: number, maximum = Number.MAX_SAFE_INTEGER) {
    return Integer.withConstraint(number => number >= minimum && number <= maximum)
}


function FixedSizeArray<T extends Runtype>(element: T, size: number) {
    return Array_(element).withConstraint(array => array.length === size)
}


function NonEmptyArray<T extends Runtype>(element: T) {
    return Array_(element).withConstraint(array => array.length > 0)
}


const NonEmptyString = String.withConstraint(string => string.length !== 0)


const PositiveInteger = IntegerRange(0)


const V1PlaceholderAlbum = Record_({
    placeholder: Literal(true),
    id: PositiveInteger
})


const SearchAlbum = Record_({
    name: NonEmptyString,
    url: NonEmptyString
})


const V1NamedAlbum = SearchAlbum.And(
    Record_({
        placeholder: Literal(false),
        id: PositiveInteger
    })
)
const V2NamedAlbum = SearchAlbum.And(
    Record_({
        id: PositiveInteger
    })
)
export const NamedAlbum = V2NamedAlbum


const V1Album = V1NamedAlbum.Or(V1PlaceholderAlbum)
const V2Album = V2NamedAlbum.Or(PositiveInteger)
const Album = V2Album


const ChartShape = Union(
    Record_({
        tag: Literal('Top'),
        size: Union(
            Literal(40),
            Literal(42),
            Literal(100)
        )
    }),
    Record_({
        tag: Literal('Collage')
    })
)


const V1Chart = Record_({
    name: NonEmptyString,
    albums: FixedSizeArray(V1Album, CHART_ALBUMS_COUNT),
    shape: ChartShape,
    rowsX: IntegerRange(1, MAX_COLLAGE_ROWS_X),
    rowsY: IntegerRange(1, MAX_COLLAGE_ROWS_Y)
})
const V2Chart = Record_({
    ...V1Chart.fields,
    albums: FixedSizeArray(Album, CHART_ALBUMS_COUNT)
})
const Chart = V2Chart


const SearchState = Union(
    Record_({
        tag: Literal('Waiting')
    }),
    Record_({
        tag: Literal('Loading'),
        controller: InstanceOf(AbortController)
    }),
    Record_({
        tag: Literal('Complete'),
        albums: NonEmptyArray(SearchAlbum)
    }),
    Record_({
        tag: Literal('Error'),
        message: String
    })
).And(
    Record_({
        query: String
    })
)


const ScreenshotState = Record_({
    loading: Boolean,
    scale: IntegerRange(1, MAX_SCREENSHOT_SCALE)
})


const ExportChartAlbum = SearchAlbum.Or(Literal(EXPORT_CHART_PLACEHOLDER))

const ExportChartShape = ChartShape.alternatives[0].Or(
    ChartShape.alternatives[1].And(
        Record_({
            rowsX: V2Chart.fields.rowsX,
            rowsY: V2Chart.fields.rowsY,
        })
    )
)

const ExportChart = Record_({
    name: V2Chart.fields.name,
    albums: FixedSizeArray(ExportChartAlbum, CHART_ALBUMS_COUNT),
    shape: ExportChartShape
})


export const V1State = Record_({
    apiKey: String,
    charts: NonEmptyArray(V1Chart),
    activeChartIndex: PositiveInteger,
    search: SearchState,
    screenshot: ScreenshotState
}).And(
    Partial_({
        version: Literal(1),
        highlightedID: PositiveInteger
    })
)
export const V2State = Record_({
    ...V1State.intersectees[0].fields,
    version: Literal(2),
    charts: NonEmptyArray(V2Chart)
}).And(
    Partial_({
        highlightedID: PositiveInteger
    })
)
const V3State = Record_({
    ...V2State.intersectees[0].fields,
    version: Literal(STATE_VERSION)
}).And(
    Partial_({
        ...V2State.intersectees[1].fields,
        viewing: ExportChart
    })
)
export const State = V3State


export type NamedAlbum = Static<typeof NamedAlbum>
export type SearchAlbum = Static<typeof SearchAlbum>
export type Album = Static<typeof Album>
export type ChartShape = Static<typeof ChartShape>
export type Chart = Static<typeof Chart>
export type SearchState = Static<typeof SearchState>
export type ScreenshotState = Static<typeof ScreenshotState>
export type State = Static<typeof State>
export type V1State = Static<typeof V1State>
export type V2State = Static<typeof V2State>
export type ExportChart = Static<typeof ExportChart>
export type ExportChartShape = Static<typeof ExportChartShape>
export type ExportChartAlbum = Static<typeof ExportChartAlbum>
