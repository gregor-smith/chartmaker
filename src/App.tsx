import React from 'react'
import { css } from 'emotion'

import { useLocalStorageSideEffectReducer } from './hooks'
import { reducer } from './reducer'
import { Chart } from './components/Chart'
import { Sidebar } from './components/Sidebar'
import { createInitialState } from './state'


const rootStyle = css({
    display: 'flex',
    minHeight: '100vh',
    backgroundColor: 'black',
    color: 'white',
    fontSize: '0.875rem',
    padding: '1rem'
})


export const App = () => {
    const [ state, dispatch ] = useLocalStorageSideEffectReducer(
        'state',
        createInitialState,
        reducer
    )

    const activeChart = state.charts.find(chart =>
        chart.name === state.activeChartName
    )
    const chart = activeChart === undefined
        ? null
        : <Chart dispatch={dispatch}
            details={activeChart}
            draggedAlbumID={state.draggedAlbumID}/>

    return (
        <div className={rootStyle}>
            <Sidebar dispatch={dispatch}
                charts={state.charts}
                activeChartName={state.activeChartName}
                apiKey={state.apiKey}
                searchState={state.search}
                draggedAlbumID={state.draggedAlbumID}/>
            {chart}
        </div>
    )
}
