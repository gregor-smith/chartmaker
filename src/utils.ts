import html2canvas from 'html2canvas'

import { BACKGROUND_COLOUR } from './style'


export function readFileText(file: File): Promise<string> {
    return new Promise((resolve, reject) => {
        const reader = new FileReader()
        reader.addEventListener('error', reject)
        reader.addEventListener('load', () => resolve(reader.result as string))
        reader.readAsText(file, 'utf-8')
    })
}


export function findIndex<T>(array: ReadonlyArray<T>, predicate: (item: T) => boolean): number | null {
    const index = array.findIndex(predicate)
    return index === -1 ? null : index
}


export async function elementToDataURI(element: HTMLElement, scale: number) {
    const canvas = await html2canvas(element, {
        allowTaint: true,
        scale,
        useCORS: true,
        imageTimeout: 0,
        backgroundColor: BACKGROUND_COLOUR
    })
    return canvas.toDataURL()
}


export function jsonToDataURI(json: string): string {
    return 'data:application/json;charset=utf-8,' + encodeURIComponent(json)
}


export function downloadURI(uri: string, filename: string) {
    const link = document.createElement('a')
    link.href = uri
    link.download = filename
    link.click()
    link.remove()
}
