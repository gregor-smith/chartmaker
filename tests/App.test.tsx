import React from 'react'
import { render } from 'react-dom'

import { State } from '@/types'
import { App } from '@/App'

import { RenderContainer } from './utils'


jest.mock('@/components/Sidebar')
jest.mock('@/components/Chart')


const localStorageMock = jest.spyOn(Storage.prototype, 'getItem')
    .mockImplementation(() => {
        const state: State = {
            charts: [
                {
                    name: 'Test chart 1',
                    rowsX: 10,
                    rowsY: 10,
                    shape: { tag: 'Top', size: 40 },
                    albums: [ ...Array(10).keys() ].map(index => ({
                        placeholder: true,
                        id: index
                    }))
                },
                {
                    name: 'Test chart 2',
                    rowsX: 3,
                    rowsY: 3,
                    shape: { tag: 'Collage' },
                    albums: [ ...Array(10).keys() ].map(index => ({
                        placeholder: true,
                        id: index + 10
                    }))
                }
            ],
            apiKey: 'Test api key',
            screenshot: {
                loading: false,
                scale: 2
            },
            search: {
                tag: 'Complete',
                query: 'Test search query',
                albums: [ ...Array(5).keys() ].map(index => {
                    const id = index + 20
                    return {
                        placeholder: false,
                        id,
                        name: `Test album ${id}`,
                        url: `https://test.com/${id}`
                    }
                })
            },
            activeChartIndex: 1,
            albumIDCounter: 205
        }
        return JSON.stringify(state)
    })
afterEach(() => localStorageMock.mockReset())


const container = new RenderContainer()


test('renders sidebar and chart', () => {
    render(
        <App/>,
        container.element
    )

    expect(container.element).toMatchSnapshot()
})
