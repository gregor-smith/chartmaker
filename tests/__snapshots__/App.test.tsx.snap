// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders sidebar and chart 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: start;
  min-height: 100vh;
  min-width: -webkit-max-content;
  min-width: -moz-max-content;
  min-width: max-content;
  background: black;
  color: white;
  font-size: 0.875rem;
  padding: 1rem;
}

<div
  id="test-render-container"
>
  <div
    class="emotion-0"
  >
    <div
      class="mock-sidebar"
    >
      Charts: [
  {
    "name": "Untitled chart",
    "albums": [
      {
        "placeholder": true,
        "id": 1
      },
      {
        "placeholder": true,
        "id": 2
      },
      {
        "placeholder": true,
        "id": 3
      },
      {
        "placeholder": true,
        "id": 4
      },
      {
        "placeholder": true,
        "id": 5
      },
      {
        "placeholder": true,
        "id": 6
      },
      {
        "placeholder": true,
        "id": 7
      },
      {
        "placeholder": true,
        "id": 8
      },
      {
        "placeholder": true,
        "id": 9
      },
      {
        "placeholder": true,
        "id": 10
      },
      {
        "placeholder": true,
        "id": 11
      },
      {
        "placeholder": true,
        "id": 12
      },
      {
        "placeholder": true,
        "id": 13
      },
      {
        "placeholder": true,
        "id": 14
      },
      {
        "placeholder": true,
        "id": 15
      },
      {
        "placeholder": true,
        "id": 16
      },
      {
        "placeholder": true,
        "id": 17
      },
      {
        "placeholder": true,
        "id": 18
      },
      {
        "placeholder": true,
        "id": 19
      },
      {
        "placeholder": true,
        "id": 20
      },
      {
        "placeholder": true,
        "id": 21
      },
      {
        "placeholder": true,
        "id": 22
      },
      {
        "placeholder": true,
        "id": 23
      },
      {
        "placeholder": true,
        "id": 24
      },
      {
        "placeholder": true,
        "id": 25
      },
      {
        "placeholder": true,
        "id": 26
      },
      {
        "placeholder": true,
        "id": 27
      },
      {
        "placeholder": true,
        "id": 28
      },
      {
        "placeholder": true,
        "id": 29
      },
      {
        "placeholder": true,
        "id": 30
      },
      {
        "placeholder": true,
        "id": 31
      },
      {
        "placeholder": true,
        "id": 32
      },
      {
        "placeholder": true,
        "id": 33
      },
      {
        "placeholder": true,
        "id": 34
      },
      {
        "placeholder": true,
        "id": 35
      },
      {
        "placeholder": true,
        "id": 36
      },
      {
        "placeholder": true,
        "id": 37
      },
      {
        "placeholder": true,
        "id": 38
      },
      {
        "placeholder": true,
        "id": 39
      },
      {
        "placeholder": true,
        "id": 40
      },
      {
        "placeholder": true,
        "id": 41
      },
      {
        "placeholder": true,
        "id": 42
      },
      {
        "placeholder": true,
        "id": 43
      },
      {
        "placeholder": true,
        "id": 44
      },
      {
        "placeholder": true,
        "id": 45
      },
      {
        "placeholder": true,
        "id": 46
      },
      {
        "placeholder": true,
        "id": 47
      },
      {
        "placeholder": true,
        "id": 48
      },
      {
        "placeholder": true,
        "id": 49
      },
      {
        "placeholder": true,
        "id": 50
      },
      {
        "placeholder": true,
        "id": 51
      },
      {
        "placeholder": true,
        "id": 52
      },
      {
        "placeholder": true,
        "id": 53
      },
      {
        "placeholder": true,
        "id": 54
      },
      {
        "placeholder": true,
        "id": 55
      },
      {
        "placeholder": true,
        "id": 56
      },
      {
        "placeholder": true,
        "id": 57
      },
      {
        "placeholder": true,
        "id": 58
      },
      {
        "placeholder": true,
        "id": 59
      },
      {
        "placeholder": true,
        "id": 60
      },
      {
        "placeholder": true,
        "id": 61
      },
      {
        "placeholder": true,
        "id": 62
      },
      {
        "placeholder": true,
        "id": 63
      },
      {
        "placeholder": true,
        "id": 64
      },
      {
        "placeholder": true,
        "id": 65
      },
      {
        "placeholder": true,
        "id": 66
      },
      {
        "placeholder": true,
        "id": 67
      },
      {
        "placeholder": true,
        "id": 68
      },
      {
        "placeholder": true,
        "id": 69
      },
      {
        "placeholder": true,
        "id": 70
      },
      {
        "placeholder": true,
        "id": 71
      },
      {
        "placeholder": true,
        "id": 72
      },
      {
        "placeholder": true,
        "id": 73
      },
      {
        "placeholder": true,
        "id": 74
      },
      {
        "placeholder": true,
        "id": 75
      },
      {
        "placeholder": true,
        "id": 76
      },
      {
        "placeholder": true,
        "id": 77
      },
      {
        "placeholder": true,
        "id": 78
      },
      {
        "placeholder": true,
        "id": 79
      },
      {
        "placeholder": true,
        "id": 80
      },
      {
        "placeholder": true,
        "id": 81
      },
      {
        "placeholder": true,
        "id": 82
      },
      {
        "placeholder": true,
        "id": 83
      },
      {
        "placeholder": true,
        "id": 84
      },
      {
        "placeholder": true,
        "id": 85
      },
      {
        "placeholder": true,
        "id": 86
      },
      {
        "placeholder": true,
        "id": 87
      },
      {
        "placeholder": true,
        "id": 88
      },
      {
        "placeholder": true,
        "id": 89
      },
      {
        "placeholder": true,
        "id": 90
      },
      {
        "placeholder": true,
        "id": 91
      },
      {
        "placeholder": true,
        "id": 92
      },
      {
        "placeholder": true,
        "id": 93
      },
      {
        "placeholder": true,
        "id": 94
      },
      {
        "placeholder": true,
        "id": 95
      },
      {
        "placeholder": true,
        "id": 96
      },
      {
        "placeholder": true,
        "id": 97
      },
      {
        "placeholder": true,
        "id": 98
      },
      {
        "placeholder": true,
        "id": 99
      },
      {
        "placeholder": true,
        "id": 100
      }
    ],
    "shape": {
      "tag": "Top",
      "size": 40
    },
    "rowsX": 5,
    "rowsY": 5
  }
]
      Screenshot state: {
  "loading": false,
  "scale": 1
}
      Search state: {
  "tag": "Waiting",
  "query": ""
}
      API key: 
      Active chart index: 0
    </div>
    <div
      class="mock-chart"
    >
      Albums: [
  {
    "placeholder": true,
    "id": 1
  },
  {
    "placeholder": true,
    "id": 2
  },
  {
    "placeholder": true,
    "id": 3
  },
  {
    "placeholder": true,
    "id": 4
  },
  {
    "placeholder": true,
    "id": 5
  },
  {
    "placeholder": true,
    "id": 6
  },
  {
    "placeholder": true,
    "id": 7
  },
  {
    "placeholder": true,
    "id": 8
  },
  {
    "placeholder": true,
    "id": 9
  },
  {
    "placeholder": true,
    "id": 10
  },
  {
    "placeholder": true,
    "id": 11
  },
  {
    "placeholder": true,
    "id": 12
  },
  {
    "placeholder": true,
    "id": 13
  },
  {
    "placeholder": true,
    "id": 14
  },
  {
    "placeholder": true,
    "id": 15
  },
  {
    "placeholder": true,
    "id": 16
  },
  {
    "placeholder": true,
    "id": 17
  },
  {
    "placeholder": true,
    "id": 18
  },
  {
    "placeholder": true,
    "id": 19
  },
  {
    "placeholder": true,
    "id": 20
  },
  {
    "placeholder": true,
    "id": 21
  },
  {
    "placeholder": true,
    "id": 22
  },
  {
    "placeholder": true,
    "id": 23
  },
  {
    "placeholder": true,
    "id": 24
  },
  {
    "placeholder": true,
    "id": 25
  },
  {
    "placeholder": true,
    "id": 26
  },
  {
    "placeholder": true,
    "id": 27
  },
  {
    "placeholder": true,
    "id": 28
  },
  {
    "placeholder": true,
    "id": 29
  },
  {
    "placeholder": true,
    "id": 30
  },
  {
    "placeholder": true,
    "id": 31
  },
  {
    "placeholder": true,
    "id": 32
  },
  {
    "placeholder": true,
    "id": 33
  },
  {
    "placeholder": true,
    "id": 34
  },
  {
    "placeholder": true,
    "id": 35
  },
  {
    "placeholder": true,
    "id": 36
  },
  {
    "placeholder": true,
    "id": 37
  },
  {
    "placeholder": true,
    "id": 38
  },
  {
    "placeholder": true,
    "id": 39
  },
  {
    "placeholder": true,
    "id": 40
  },
  {
    "placeholder": true,
    "id": 41
  },
  {
    "placeholder": true,
    "id": 42
  },
  {
    "placeholder": true,
    "id": 43
  },
  {
    "placeholder": true,
    "id": 44
  },
  {
    "placeholder": true,
    "id": 45
  },
  {
    "placeholder": true,
    "id": 46
  },
  {
    "placeholder": true,
    "id": 47
  },
  {
    "placeholder": true,
    "id": 48
  },
  {
    "placeholder": true,
    "id": 49
  },
  {
    "placeholder": true,
    "id": 50
  },
  {
    "placeholder": true,
    "id": 51
  },
  {
    "placeholder": true,
    "id": 52
  },
  {
    "placeholder": true,
    "id": 53
  },
  {
    "placeholder": true,
    "id": 54
  },
  {
    "placeholder": true,
    "id": 55
  },
  {
    "placeholder": true,
    "id": 56
  },
  {
    "placeholder": true,
    "id": 57
  },
  {
    "placeholder": true,
    "id": 58
  },
  {
    "placeholder": true,
    "id": 59
  },
  {
    "placeholder": true,
    "id": 60
  },
  {
    "placeholder": true,
    "id": 61
  },
  {
    "placeholder": true,
    "id": 62
  },
  {
    "placeholder": true,
    "id": 63
  },
  {
    "placeholder": true,
    "id": 64
  },
  {
    "placeholder": true,
    "id": 65
  },
  {
    "placeholder": true,
    "id": 66
  },
  {
    "placeholder": true,
    "id": 67
  },
  {
    "placeholder": true,
    "id": 68
  },
  {
    "placeholder": true,
    "id": 69
  },
  {
    "placeholder": true,
    "id": 70
  },
  {
    "placeholder": true,
    "id": 71
  },
  {
    "placeholder": true,
    "id": 72
  },
  {
    "placeholder": true,
    "id": 73
  },
  {
    "placeholder": true,
    "id": 74
  },
  {
    "placeholder": true,
    "id": 75
  },
  {
    "placeholder": true,
    "id": 76
  },
  {
    "placeholder": true,
    "id": 77
  },
  {
    "placeholder": true,
    "id": 78
  },
  {
    "placeholder": true,
    "id": 79
  },
  {
    "placeholder": true,
    "id": 80
  },
  {
    "placeholder": true,
    "id": 81
  },
  {
    "placeholder": true,
    "id": 82
  },
  {
    "placeholder": true,
    "id": 83
  },
  {
    "placeholder": true,
    "id": 84
  },
  {
    "placeholder": true,
    "id": 85
  },
  {
    "placeholder": true,
    "id": 86
  },
  {
    "placeholder": true,
    "id": 87
  },
  {
    "placeholder": true,
    "id": 88
  },
  {
    "placeholder": true,
    "id": 89
  },
  {
    "placeholder": true,
    "id": 90
  },
  {
    "placeholder": true,
    "id": 91
  },
  {
    "placeholder": true,
    "id": 92
  },
  {
    "placeholder": true,
    "id": 93
  },
  {
    "placeholder": true,
    "id": 94
  },
  {
    "placeholder": true,
    "id": 95
  },
  {
    "placeholder": true,
    "id": 96
  },
  {
    "placeholder": true,
    "id": 97
  },
  {
    "placeholder": true,
    "id": 98
  },
  {
    "placeholder": true,
    "id": 99
  },
  {
    "placeholder": true,
    "id": 100
  }
]
      Name: Untitled chart
      Rows X: 5
      Rows Y: 5
      Shape: {"tag":"Top","size":40}
    </div>
  </div>
</div>
`;
